package/MDAnalysis/lib/pkdtree.py

```diff
@@ -0,0 +1,172 @@
+
+        Parameters
+        ----------
+        center_point: NumPy.array
+          Coordinates of the query center point
+        radius: float
+          Maximum distance from center in search for neighbors
+
+        Returns
+        ------
+        :class:`List`
+          wrapped center point and its relevant images
+        """
+        wrapped_c = apply_PBC(center_point.reshape(1, 3), self.box)[0]
+        extents = self.box/2.0
+        extents = np.where(extents > radius, radius, extents)
+        # displacements are vectors that we add to wrapped_c to
+        # generate images "up" or "down" the central cell along
+        # the axis that we happen to be looking.
+        #
+        # TO-DO: extend to a triclinic box by finding distance from wrapped_c
+        # to each of the planes enclosing the central cell.
+        displacements = list()
+        for i in range(self.dim):
+            displacement = np.zeros(self.dim)
+            extent = extents[i]
+            if extent > 0.0:
+                if self.box[i] - wrapped_c[i] < extent:
+                    displacement[i] = -self.box[i]  # "lower" image
+                    displacements.append(displacement)
+                elif wrapped_c[i] < extent:
+                    displacement[i] = self.box[i]  # "upper" image
+                    displacements.append(displacement)
+        # If we have displacements along more than one axis, we have
+        # to combine them. This happens when wrapped_center is close
+        # to any edge or vertex of the central cell.
+        # face, n_displacements==1; no combination
+        # edge, n_displacements==2; combinations produce one extra displacement
+        # vertex, n_displacements==3; five extra displacements
+        n_displacements = len(displacements)
+        if n_displacements > 1:
+            for start in range(n_displacements - 1, -1, -1):
+                for i in range(start+1, len(displacements)):
+                    displacements.append(displacements[start]+displacements[i])
+        return [wrapped_c, ] + [wrapped_c + d for d in displacements]
+
+    def search(self, center, radius):
+        """Search all points within radius of center and its periodic images.
+        Wrapping of center coordinates is enforced to enable comparison to
+        wrapped coordinates of points in the tree.
+
+        Parameter
+        ---------
+        center: NumPy.array
+          origin around which to search for neighbors
+        radius: float
+          maximum distance around which to search for neighbors. The search
+          radius is half the smallest periodicity if radius exceeds this value
+        """
+        if not self.built:
+            raise RuntimeError('Unbuilt tree. Run tree.set_coords first')
+        if center.shape != (self.dim,):
+            raise ValueError('Expected a ({},) NumPy array'.format(self.dim))
+        self._indices = None  # clear previous search
```

Comment:
iirc, calling `np.append` repeatedly is slower than using a list.  I think it'd be quicker and clearer to do `self._indices = []` then just keep extending the list `if new_indices is not None:`.  Then only turn it into a numpy array at the end where you sort and unique it

Reply:
Does that matter when we maximally call it 4 times? Or maybe 9 times when the radius is larger then a box length.

Reply:
Ah I didn't check how many times this is called.  [either way this line isn't really clear to me](https://github.com/MDAnalysis/mdanalysis/pull/1660/files#diff-08b446f64bc4574c8b9b7714e83ecfa1R166).  I'd rather initialise an empty array here and append to it, rather than None or an array

Reply:
Still, I'd rather set the example and not have that in the code base. Each `np.append` call wastefully copies the entire array...

Reply:
I can use a python list, but since `self.kdt.get_indices()` returns a numpy array, I will have to cast to a python list before extending `self._indices`. Still, I guess it will yield some speed gain over `np.append`

Reply:
@jmborr I still think it'll be quicker to create a list of whatever `get_indices()` returns and only make the final numpy array once.

Reply:
Indeed, first use a list to collect all `new_indices` that aren't `None`, then use `np.concatenate` on it, if not empty.